File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        queue<pair<int, int>> q;
        vector<vector<int>> dir{{0, -1},{0, 1},{1, 0},{-1, 0}};
        int numberIslands = 0, m = grid.size(), n = grid[0].size(), size = 0;
        for(auto i=0; i<m; i++){
            for(auto j=0; j<n; j++){
                if(grid[i][j] == '1'){
                    numberIslands++;
                    q.push({i, j});
                    grid[i][j] = '2';
                    while(q.size() > 0){
                        size = q.size();
                        for(auto k = 0; k<size; k++){
                            auto currentNode = q.front();
                            q.pop();
                            auto x = currentNode.first, y = currentNode.second;
                            for(auto l=0; l<4; l++){
                                auto nextI = x + dir[l][0], nextJ = y + dir[l][1];
                                if(nextI >= 0 && nextJ >=0 && nextI < m && nextJ < n && grid[nextI][nextJ] == '1'){
                                    grid[nextI][nextJ] = '2';
                                    q.push({nextI, nextJ});
                                }
                            }
                        }
                    }
                }
            }
        }
        return numberIslands;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

    def dfs(self, i: int, j: int, grid: List[List[str]]) -> None:
        if 0 <= i < len(grid) and 0 <= j < len(grid[i]) and grid[i][j] == "1":
            grid[i][j] = "0"
            for direction in self.directions:
                self.dfs(i + direction[0], j + direction[1], grid)

    def numIslands(self, grid: List[List[str]]) -> int:
        result = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    self.dfs(i, j, grid)
                    result += 1
        
        return result